src/modelterms.jl

```diff
@@ -198,14 +198,14 @@ end
 function Base.sparse(A::VectorFactorReTerm{T,V,R,S}) where {T,V,R,S}
     n = size(A, 1)
     colind = Matrix{Int32}(S, n)
-    rr = A.f.refs
+    ord_lookup = A.f.pool.invindex
     @inbounds for j in 1:n
-        offset = (rr[j] - 1) * S
+        offset = (ord_lookup[A.f[j]] - 1) * S
```

Comment:
I don't understand why you do this. It doesn't make a lot of sense to extract the value and then lookup its reference code inside the pool. If I'm not mistaken that's strictly equivalent to `A.f.refs[j]`. Doesn't the solution based on `order` work? Or maybe you need the inverse of `order`?

Reply:
Good question. The solution based on order does not work, it broke the tests because it was looking up the value. Have a look at the object created by the pool method of a `CategoricalArray`, and what `order()` on that gives you back (which is a list of the refs). We needed a dictionary to look that up from the values of `A.f`.

From what I can tell, the difference between this lookup using the `ref` is that this respects the full set of levels (whereas refs will only use the levels seen in the data). By going this new way, you allow a user to specify levels on the data programmatically.

Reply:
No, I don't think that's the problem here. You're iterating over the entries in `A.f` in this loop, so only values present in the data are considered. And `order` returns the ordering of all `levels`, not only of those present in the data. Can you explain exactly what the problem with the `order`-based approach?

Reply:
What exactly do you mean by the `order`-based approach? I tried the code you posted before, and it didn't work, so I fixed it with this.

Reply:
This is the code you posted before, which is what I assume you mean

    ord = CategoricalArrays.order(A.f.pool); ord[A.f[j]]

But that simply doesn't work.

Reply:
The exact equivalent of what you're currently doing is just `A.ref[j]`. So if that's really what you need, just use that, but that's also what the old code did.

The problem is, I don't think that's what's is intended here. The approach described above returns levels in their internal storage order, not in the user-visible and user-defined order defined by `levels`. `order` allows passing from the internal codes to the user-visible codes, if that's what is intended.

Reply:
Oh, I just noted a possible explanation for the failure of the example you give in your last comment. It should be:
```julia
ord = CategoricalArrays.order(A.f.pool); ord[A.f.refs[j]]
```

Reply:
Oh yeah that fixes yours, and makes more sense.